### PR TITLE
Add E0790 as more specific variant of E0283

### LINKED ISSUE #81701: E0238 could be better for calls to associated functions on the trait itself
This code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=1afd39b6d1676f0b09b79f09120f60d9))

```rust
trait MyTrait {
    fn my_fn();
}

struct MyStruct;

impl MyTrait for MyStruct {
    fn my_fn() {}
}

fn main() {
    MyTrait::my_fn();
}
```

currently produces this error:

```text
error[E0283]: type annotations needed
  --> src/main.rs:12:5
   |
2  |     fn my_fn();
   |     ----------- required by `MyTrait::my_fn`
...
12 |     MyTrait::my_fn();
   |     ^^^^^^^^^^^^^^ cannot infer type
   |
   = note: cannot satisfy `_: MyTrait`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0283`.
```

When I recently stumbled across this error, it took me a little while to figure out what the problem was.
I was trying to call the trait function directly and not an implementation of the function.
I think this may be even more confusing for beginners.

I think the following error message would have helped me (and may help others) understand what the problem is and how to fix it much more quickly:

```text
error[E0283]: a concrete implementation is needed to call the trait function
  --> src/main.rs:12:5
   |
12 |     MyTrait::my_fn();
   |     ^^^^^^^^^^^^^^ cannot infer concrete implementation of `my_fn`
   |
   = note: cannot satisfy `_: MyTrait`
   = help: try calling `my_fn` on an implementation of `MyTrait` such as `MyStruct::my_fn`
   = help: available implementations are `MyStruct`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0283`.
```

### PR DESCRIPTION
**Pull request by aticu**
Fixes #81701

I think this should be good to go, there are only two things where I am somewhat unsure:
- Is there a better way to get the fully-qualified path for the suggestion? I tried `self.tcx.def_path_str`, but that didn't seem to always give a correct path for the context.
- Should all this be extracted into it's own method or is it fine where it is?

r? @estebank

### CODE REVIEW

File: compiler/rustc_error_codes/src/error_codes/E0789.md
Comment (GuillaumeGomez) [inline compiler/rustc_error_codes/src/error_codes/E0789.md:23]: You don't need to have a `main` function, you can just put the code directly.

File: compiler/rustc_error_codes/src/error_codes/E0789.md
Comment (GuillaumeGomez) [inline compiler/rustc_error_codes/src/error_codes/E0789.md:45]: Same.

**Review by eggyal — CHANGES_REQUESTED** [review]

File: compiler/rustc_error_codes/src/error_codes/E0283.md
Comment (eggyal) [inline compiler/rustc_error_codes/src/error_codes/E0283.md:28]: Oops, did I suggest incorrectly?  That won't parse.  You either need `<Into<u32>>::into` or `Into::<u32>::into`.  Sorry!!

**Review by eggyal — CHANGES_REQUESTED** [review]

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
The error explanation looks good to me. Let's wait for @estebank for the approval on the compiler change.

**Comment by aticu** [comment]
I've slightly changed the structure of the code now in hope to make the diff simpler, but nothing fundamentally changed in case you already started looking into this @estebank

**Comment by Dylan-DPC** [comment]
failed in [rollup ci](https://github.com/rust-lang/rust/pull/99450#issuecomment-1188636053)

**Comment by aticu** [comment]
Rebased and changed the error code to E0790.

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:2111]: I wonder if we could introduce a new `ObligationCauseCode` that stores the hir id of the `ExprKind::Path` that this obligation comes from, so we wouldn't need to use this span-based search method.

File: src/test/ui/error-codes/E0283.stderr
Comment (compiler-errors) [inline src/test/ui/error-codes/E0283.stderr:12]: This `::Impl` path is kinda funky, is there a reason we used `to_string_no_crate_verbose` instead of just `tcx.def_path_str`?

**Comment by compiler-errors** [comment]
@aticu, does this fix address #98938 as well?

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs
Comment (estebank) [reply compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:2111]: That sounds like a good idea.

File: src/test/ui/error-codes/E0283.stderr
Comment (aticu) [reply src/test/ui/error-codes/E0283.stderr:12]: I don't recall exactly anymore, but if I remember correctly I tried that and got paths as a result that didn't make sense in the context where they were suggested. But I could have made a mistake then. If you'd like I could try it again when I find the time and make a follow-up PR with that or at least find out why it didn't work in this case.

**Comment by aticu** [comment]
@compiler-errors Yes, this does fix #98938 as well, I just double checked. The resulting error message in that case is

```text
error[E0790]: cannot call associated function on trait without specifying the corresponding `impl` type
 --> test.rs:6:5
  |
2 |     fn bar() {}
  |     ----------- `Foo::bar` defined here
...
6 |     Foo::bar();
  |     ^^^^^^^^ cannot call associated function of trait
```

File: src/test/ui/error-codes/E0283.stderr
Comment (compiler-errors) [reply src/test/ui/error-codes/E0283.stderr:12]: Yeah, either put up a follow up or I can also look into it.

Comment (davidkna) [reply src/test/ui/error-codes/E0283.stderr:12]: (I was involved in the development of this PR)
I think it was along the lines of it returning `Impl` instead of the context-sensitive value `inner::Impl` in the case of nested modules.

**Pull request merged**
